design-proposals/isolate-emulator-thread-with-smt-fix/isolate-emulator-thread-with-smt-fix.md

```diff
@@ -54,13 +54,75 @@ Fix the actual scheduling of the virt-launcher to a node that will reject it.
 This bug addresses all KubeVirt users running VMs with isolateEmulatorThread on nodes with hyperthreading enabled - specifically DPDK and Realtime users.
 
 ## User Stories
-A user should be able to run VM with dedicatedCPUPlancement and isolateEmulatorThread enabled on a node with hyperthreading enabled, regardless of the amount of guest CPUs requested.
+* As a cluster admin, I want to set the cluster-wide IsolatedEmulatorThread behavior for Kubevirt by setting a knob on HCO.
+* as a VM owner, I should be able to run VM with dedicatedCPUPlancement and isolateEmulatorThread enabled on a node with hyperthreading enabled, regardless of the amount of guest CPUs requested.
 
 ## Repos
 - [kubevirt/kubevirt](https://github.com/kubevirt/kubevirt)
+- [kubevirt/hyperconverged-cluster-operator](https://github.com/kubevirt/hyperconverged-cluster-operator)
 
 # Bug fix design proposal
-The bug fix proposal introduces an alpha annotation `kubevirt.io/CPUManagerPolicyBetaOptions:full-pcpus-only`:
+The bug fix proposal introduces an alpha annotation to the kubevirt CR: `kubevirt.io/EmulatorThreadEvenParity`:
+```
+apiVersion: kubevirt.io/v1
+kind: KubeVirt
+metadata:
+  annotations:
+    kubevirt.io/EmulatorThreadEvenParity:
+spec:
+  configuration:
+    developerConfiguration:
+      featureGates:
+      - EmulatorThreadEvenParity
+```
+This annotation will be supported only when the `EmulatorThreadEvenParity` featureGate is enabled.
+
+## Virt-API:
+When the following conditions are met:
+1. kubevirt CR has the annotation.
+2. kubevirt CR enabled the feature gate.
+3. VM is configured with `isolateEmulatorThread: true`
+
+then the VMI mutating-webhook will
+1. Copy the annotation from the kubevirt CR to the VMI annotations. This is done in order for it to be used on the virt-launcher, when setting the housekeeping cgroup. Doing so avoids adding API on the virt-handler to virt-launcher grpc protocol.
+
```

Comment:
There is an additional step, already implemented and merged on main:
The VMI creation admitter should allow the VMI to have the new annotation only if:
1. The FG is enabled.
2. The annotation on the Kubevirt CR is set.

Reply:
I think @vladikr was against it. am I wrong?

Reply:
Please see: https://github.com/kubevirt/kubevirt/commit/bd8987871c2dc1b42dcece8a869c24e4e02c8d85

Reply:
I don't think it's the same..
In our case, the user (i.e. the VM owner) is not the one who's setting the annotation on the VMI.. it's an internal thing we do.

Reply:
It is still possible for a user to manually set this annotation.
And IMO we should make sure that it is only used when the FG is enabled.

Reply:
talked offline, agreed the VMI create-admitter validation webhook should only allow the VMI if:
1. The feature gate is enabled on the Kubevirt CR
2. VMI is configured with `isolateEmulatorThread: true`